README.md
hashID
======

Identify the different types of hashes used to encrypt data and especially passwords.

This tool replaces [hash-identifier](http://code.google.com/p/hash-identifier/ "hash-identifier"), which is outdated!
 
hashID is a tool written in Python 3.x which supports the identification of over 200 unique hash types using regular expressions.
A full list of supported hashes is found [here](hashinfo.xlsx "hashinfo.xlsx").    
It is able to identify a single hash, parse a file or read files in a directory and identify the hashes within them.    
hashID is also capable of including the corresponding [hashcat](https://hashcat.net/oclhashcat/ "hashcat") mode in its output.    
Altough hashID is written in Python 3.x it should also work with Python 2.7.

There is also a [nodejs](https://github.com/psypanda/hashID/tree/js "hashID-nodejs") version of hashID available which is easily set up to provide online hash identification.  

*Note: When identifying a hash on *nix operating systems use single quotes to prevent interpolation*

Install
------
```console
$ sudo apt-get install python3 git
$ git clone https://github.com/psypanda/hashid.git
$ cd hashid
$ sudo install -g 0 -o 0 -m 0644 man/hashid.7 /usr/share/man/man7/
$ sudo gzip /usr/share/man/man7/hashid.7
```

Usage
------
```console
$ ./hashid.py [-a] [-m] [--help] [--version] INPUT
```

| Parameter                      | Description                                         |
| :----------------------------- | :-------------------------------------------------- |
| input                          | string or filename to analyze                       |
| -a, --all                      | list all hash algorithms including salted passwords |
| -m, --mode                     | include corresponding hashcat mode in output        |
| --help                         | show this help message and exit                     |
| --version                      | show program's version number and exit              |


Screenshot
------
```console
$ ./hashid.py $P$8ohUJ.1sdFw09/bMaAQPTGDNi2BIUt1
Analyzing '$P$8ohUJ.1sdFw09/bMaAQPTGDNi2BIUt1'
[+] Wordpress ≥ v2.6.2
[+] Joomla ≥ v2.5.18
[+] PHPass' Portable Hash

$ ./hashid.py -m ecf076ce9d6ed3624a9332112b1cd67b236fdd11:17782686
Analyzing 'ecf076ce9d6ed3624a9332112b1cd67b236fdd11:17782686'
[+] Redmine Project Management Web App [Hashcat Mode: 7600]
[+] SMF ≥ v1.1 [Hashcat Mode: 121]

$ ./hashid.py hashes.txt
--File 'hashes.txt'--
Analyzing '*85ADE5DDF71E348162894C71D73324C043838751'
[+] MySQL5.x
[+] MySQL4.1
Analyzing '$2a$08$VPzNKPAY60FsAbnq.c.h5.XTCZtC1z.j3hnlDFGImN9FcpfR1QnLq'
[+] Blowfish(OpenBSD)
[+] Woltlab Burning Board 4.x
[+] BCrypt
--End of file 'hashes.txt'--

$ ./hashid.py folder/*.txt
--File 'hashes.txt'--
Analyzing '*85ADE5DDF71E348162894C71D73324C043838751'
[+] MySQL5.x
[+] MySQL4.1
Analyzing '$2a$08$VPzNKPAY60FsAbnq.c.h5.XTCZtC1z.j3hnlDFGImN9FcpfR1QnLq'
[+] Blowfish(OpenBSD)
[+] Woltlab Burning Board 4.x
[+] BCrypt
--End of file 'hashes.txt'--
--File 'hashes.txt'--
Analyzing '{smd5}01234567$yOImZPvBC8dg1HjGYfH7j.'
[+] AIX(smd5)
Analyzing 'crypt1:fnd+8xl+U1E=:Wc30H8MPgAc='
[+] Clavister Secure Gateway
--End of file 'hashes.txt'--
```

Contribute
------
Contributing to this project can be done in various ways:
* Add currently unsupported hash types
* Change/Fix/Enhance existing regular expression
* Provide reading resources on specific hash types in [hashinfo.xlsx](hashinfo.xlsx)

Credits
------
* Thanks to [sigkill](https://github.com/sigkill-rcode) who helped numerous times fixing and optimizing the code
* Thanks to [kmulvey](https://github.com/kmulvey) for supplying a [nodejs](https://github.com/psypanda/hashID/tree/js) version of hashID
* Thanks to [bburky](https://github.com/bburky) for extending hashIDs functionality and numerous ideas to improve it

Resources
------
* http://pythonhosted.org/passlib/index.html
* http://wiki.insidepro.com/index.php/Algorithms
* http://openwall.info/wiki/john
* http://openwall.info/wiki/john/sample-hashes
* http://hashcat.net/wiki/doku.php?id=example_hashes


CHANGELOG
### Changelog

v2.9.0
- added support for piped input (thanks Max)
- added SHA1($salt.$pass.$salt) hash format
- Django(PBKDF2-HMAC-SHA256) hashcat mode added

v2.8.5
- added Cisco Type 8 & 9 hash format
- added Android FDE ≤ 4.3 hash format
- added Microsoft Office ≤ 2003 hash format
- added Microsoft Office 2007 hash format
- added Microsoft Office 2010 hash format
- added Microsoft Office 2013 hash format
- RAdmin v2.x hashcat mode added

v2.8.4
- added Lotus Notes/Domino 8 hash format
- added scrypt hash format
- added support for raw passwd & shadow files without preprocessing
- renamed Oracle 11g to Oracle 11g/12c
- renamed Samsung Android Password/PIN to Android PIN

v2.8.3
- removed unnecessary escaping of literal characters in regexes
- reworked man page and added license notice
- fixed some length bugs mainly in the base64 hash formats
- added Skype hash format

v2.8.2
- massive speed improvements
- added man page for hashID
- added PBKDF2(Atlassian) hash format
- added PostgreSQL MD5 hash format

v2.8.1
- added PeopleSoft hash format
- added Django(DES Crypt Wrapper) hash format
- added Django(PBKDF2-HMAC-SHA256) hash format
- added Django(PBKDF2-HMAC-SHA1) hash format
- added Django(BCrypt) hash format
- added Django(MD5) hash format
- Woltlab Burning Board 3.x regex fixes
- renamed MaNGOS CMS to MangosWeb Enhanced CMS
- renamed bcrypt to BCrypt
- renamed Django CMS(SHA-1) to Django(SHA-1)
- renamed Django CMS(SHA-256) to Django(SHA-256)
- renamed Django CMS(SHA-384) to Django(SHA-384)
- MangosWeb Enhanced CMS flagged as extended
- RAdmin v2.x flagged as extended
- ZipMonster flagged as extended
- Ventrilo flagged as extended

v2.8.0
- added new parameter "-a / --all"
- PEP8 optimization
- renamed xAuth to Minecraft(xAuth)

v2.7.0
- major rewrite
- dropped obsolete "-f / --file" and "-d / --dir" parameter
- all output defaults to stdout now

v2.6.7
- small cosmetic changes
- MediaWiki hashcat mode added

v2.6.6
- added 1Password(Agile Keychain) hash format
- added 1Password(Cloud Keychain) hash format
- added IKE-PSK MD5 hash format
- added IKE-PSK SHA1 hash format
- 1Password(Agile Keychain) hashcat mode added
- 1Password(Cloud Keychain) hashcat mode added
- IKE-PSK MD5 hashcat mode added
- IKE-PSK SHA1 hashcat mode added

v2.6.5
- hashID is usable as a python module now
- PHPS regex changes
- Kerberos 5 AS-REQ Pre-Auth regex changes
    
v2.6.4
- added PHPS hash format
- PHPS hashcat mode added

v2.6.3
- added Lotus Notes/Domino 5 hash format
- DNSSEC(NSEC3) regex fixes
- renamed Lotus Domino to Lotus Notes/Domino 6
- Lotus Notes/Domino 5 hashcat mode added
- Lotus Notes/Domino 6 hashcat mode added

v2.6.2
- added JTR formats to hashinfo.xlsx
- Juniper Netscreen/SSG(ScreenOS) regex changes
- NTLM regex changes

v2.6.1
- additional mimetype check on file input
- added MSSQL(2014) hash format
- renamed Blowfish(Eggdrop) to Eggdrop IRC Bot

v2.6.0
- added directory analyze using parameter "-d / --dir"
- more error checking on user input
- vBulletin < v3.8.5 regex fixes

v2.5.0
- added CRC-24 hash format
- added Joomla ≥ 2.5.18 hash format
- added Woltlab Burning Board 4.x hash format
- GRUB 2 regex changes
- BCrypt(SHA-256) regex changes
- OSX v10.8 regex changes
- OSX v10.9 regex changes
- DNSSEC(NSEC3) regex changes
- Minecraft(AuthMe Reloaded) regex changes
- renamed Wordpress 2.6.2 to Wordpress v2.6.2
- renamed Joomla to Joomla < 2.5.18
- renamed Drupal7 to Drupal ≥ v7.x
- renamed  Burning Board 3.x to Woltlab Burning Board 3.x
- hMailServer hashcat mode added

v2.4.5
- added Fairly Secure Hashed Password hash format
- SAP CODVN B (BCODE) regex changes
- SAP CODVN F/G (PASSCODE) regex changes
- renamed Cisco-IOS(SHA256) to Cisco-IOS(SHA-256)
    
v2.4.4
- added xAuth hash format
- added PBKDF2(Generic) hash format
- added PBKDF2(Cryptacular) hash format
- added PBKDF2(Dwayne Litzenberger) hash format

v2.4.3
-  added hMailServer hash format
-  added MediaWiki hash format
-  vBulletin ≥ v3.8.5 regex changes
-  Cisco-ASA(MD5) regex changes
-  renamed BCrypt(SHA256) to BCrypt(SHA-256)
    
v2.4.2
- added SHA-1 Crypt hash format
- renamed MD5(Sun) to Sun MD5 Crypt
- renamed SHA-256(Unix) to SHA-256 Crypt
- renamed SHA-512(Unix) to SHA-512 Crypt
- renamed MD5(Unix) to MD5 Crypt

v2.4.1
- added NTHash(FreeBSD Variant) hash format
- Domain Cached Credentials regex changes
- Domain Cached Credentials 2 regex changes
- NetNTLMv1-VANILLA / NetNTLMv1+ESS regex changes
- NetNTLMv2 regex changes

v2.4.0
- added Hashcat Mode output using parameter "-hc / --hashcat"
- added RACF hash format
- added Double MD5 hash format
- added Double SHA-1 hash format
- added md5apr1 hash format
- added bcrypt hash format
- NetNTLMv1-VANILLA / NetNTLMv1+ESS regex changes
- NetNTLMv2 regex changes
    
v2.3.6
- added Burning Board 3.x hash format
- added IPMI2 RAKP HMAC-SHA1 hash format
- added Lastpass hash format
- added Cisco-ASA(MD5) hash format
- added DNSSEC(NSEC3) hash format
- VNC regex changes
- renamed Keccak to SHA3

v2.3.5
- added Wordpress v2.6.0/2.6.1 hash format
- MyBB ≥ v1.2+ regex changes
- renamed MD5(phpBB) to phpBB 3.x
- renamed MD5(Wordpress) to Wordpress ≥ v2.6.2
- renamed MD5(ZipMonster) to ZipMonster
- renamed MD5(MaNGOS) to MaNGOS CMS
- renamed MD5(MaNGOS2) to MaNGOS CMS v2
- renamed SHA1(LinkedIn)to LinkedIn
- renamed SHA-1(Django) to Django CMS(SHA-1)
- renamed SHA-256(Django) to Django CMS(SHA-256)
- renamed SHA-384(Django) to Django CMS(SHA-384)

js/hashid.js
// nodejs version of hashID
// This file is part of hashID.
//
// hashID is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// hashID is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with hashID. If not, see <http://www.gnu.org/licenses/>.

(function(exports){
    'use strict';
    
    // INPUT: a single line hash
    // OUTPUT: an array of matches
    // ERROR: an error is thrown if there are no matches
    exports.identifyHash = function(phash){
        //trim possible whitespace
        phash = phash.trim();
        //set regex and algorithms
        var hash_arr = [
            {regex: /^[a-f0-9]{4}$/i, hashes: ["CRC-16","CRC-16-CCITT","FCS-16"]},
            {regex: /^[a-f0-9]{6}$/i, hashes: ["CRC-24"]},
            {regex: /^[a-f0-9]{8}$/i, hashes: ["Adler-32","CRC-32","CRC-32B","FCS-32","GHash-32-3","GHash-32-5","FNV-132","Fletcher-32","Joaat","ELF-32","XOR-32"]},
            {regex: /^\+[a-z0-9\/\.]{12}$/i, hashes: ["Blowfish(Eggdrop)"]},
            {regex: /^[a-z0-9\/\.]{13}$/i, hashes: ["DES(Unix)","Traditional DES","DEScrypt"]},
            {regex: /^[a-f0-9]{16}$/i, hashes: ["MySQL323","DES(Oracle)","Half MD5","Oracle 7-10g","FNV-164","CRC-64"]},
            {regex: /^[a-z0-9\/\.]{16}$/i, hashes: ["Cisco-PIX(MD5)"]},
            {regex: /^\([a-z0-9\+\/]{20}\)$/i, hashes: ["Lotus Notes/Domino 6"]},
            {regex: /^_[a-z0-9\/\.]{19}$/i, hashes: ["BSDi Crypt"]},
            {regex: /^[a-f0-9]{24}$/i, hashes: ["CRC-96(ZIP)"]},
            {regex: /^[a-z0-9\/\.]{24}$/i, hashes: ["Crypt16"]},
            {regex: /^[a-f0-9]{32}$/i, hashes: ["MD5","MD4","MD2","Double MD5","NTLM","LM","RAdmin v2.x","RIPEMD-128","Haval-128","Tiger-128","Snefru-128","ZipMonster","Skein-256(128)","Skein-512(128)","Lotus Notes/Domino 5"]},
            {regex: /^[a-f0-9]{32}(:[^\\\/\:\*\?\"<\>\|]{1,20})?$/i, hashes: ["Domain Cached Credentials", "mscash"]},
            {regex: /^(\$DCC2\$10240#[^\\\/\:\*\?\"<\>\|]{1,20}#)?[a-f0-9]{32}$/i, hashes: ["Domain Cached Credentials 2","mscash2"]},
            {regex: /^{SHA}[a-z0-9\/\+]{27}=$/i, hashes: ["SHA-1(Base64)","Netscape LDAP SHA","nsldap"]},
            {regex: /^\$1\$[a-z0-9\/.]{0,8}\$[a-z0-9\/.]{22}(:.*)?$/i, hashes: ["MD5 Crypt","Cisco-IOS(MD5)","FreeBSD MD5"]},
            {regex: /^0x[a-f0-9]{32}$/i, hashes: ["Lineage II C4"]},
            {regex: /^\$H\$[a-z0-9\/\.]{31}$/i, hashes: ["phpBB v3.x","Wordpress v2.6.0/2.6.1","PHPass' Portable Hash"]},
            {regex: /^\$P\$[a-z0-9\/\.]{31}$/i, hashes: ["Wordpress ≥ 2.6.2","Joomla ≥ v2.5.18","PHPass' Portable Hash"]},
            {regex: /^[a-f0-9]{32}:[a-z0-9]{2}$/i, hashes: ["osCommerce","xt:Commerce"]},
            {regex: /^\$apr1\$[a-z0-9\/\.]{0,8}\$[a-z0-9\/\.]{22}$/i, hashes: ["MD5(APR)","Apache MD5","md5apr1"]},
            {regex: /^{smd5}[a-z0-9\.\$]{31}$/i, hashes: ["AIX(smd5)"]},
            {regex: /^[a-f0-9]{32}:[a-f0-9]{32}$/i, hashes: ["WebEdition CMS"]},
            {regex: /^[a-f0-9]{32}:.{5}$/i, hashes: ["IP.Board v2+"]},
            {regex: /^[a-f0-9]{32}:.{8}$/i, hashes: ["MyBB ≥ v1.2+"]},
            {regex: /^[a-z0-9]{34}$/i, hashes: ["CryptoCurrency(Adress)"]},
            {regex: /^[a-f0-9]{40}$/i, hashes: ["SHA-1","Double SHA-1","MangosWeb Enhanced CMS","LinkedIn","RIPEMD-160","Haval-160","Tiger-160","HAS-160","Skein-256(160)","Skein-512(160)"]},
            {regex: /^\*[a-f0-9]{40}$/i, hashes: ["MySQL5.x","MySQL4.1"]},
            {regex: /^[a-z0-9]{43}$/i, hashes: ["Cisco-IOS(SHA-256)"]},
            {regex: /^{SSHA}([a-z0-9\+\/]{40}|[a-z0-9\+\/]{38}==)$/i, hashes: ["SSHA-1(Base64)","Netscape LDAP SSHA","nsldaps"]},
            {regex: /^[a-z0-9]{47}$/i, hashes: ["Fortigate(FortiOS)"]},
            {regex: /^[a-f0-9]{48}$/i, hashes: ["Haval-192","Tiger-192","SHA-1(Oracle)","OSX v10.4","OSX v10.5","OSX v10.6"]},
            {regex: /^[a-f0-9]{51}$/i, hashes: ["Palshop CMS"]},
            {regex: /^[a-z0-9]{51}$/i, hashes: ["CryptoCurrency(PrivateKey)"]},
            {regex: /^{ssha1}[a-z0-9\.\$]{47}$/i, hashes: ["AIX(ssha1)"]},
            {regex: /^0x0100[a-f0-9]{48}$/i, hashes: ["MSSQL(2005)","MSSQL(2008)"]},
            {regex: /^(\$md5,rounds=[0-9]+\$|\$md5\$rounds=[0-9]+\$|\$md5\$)[a-z0-9\/\.]{0,16}(\$|\$\$)[a-z0-9\/\.]{22}$/i, hashes: ["Sun MD5 Crypt"]},
            {regex: /^[a-f0-9]{56}$/i, hashes: ["SHA-224","Haval-224","SHA3-224","Skein-256(224)","Skein-512(224)"]},
            {regex: /^(\$2[axy]|\$2)\$[0-9]{0,2}?\$[a-z0-9\/\.]{53}$/i, hashes: ["Blowfish(OpenBSD)","Woltlab Burning Board 4.x","BCrypt"]},
            {regex: /^[a-f0-9]{40}:[a-f0-9]{16}$/i, hashes: ["Samsung Android Password/PIN"]},
            {regex: /^S:[a-f0-9]{60}$/i, hashes: ["Oracle 11g"]},
            {regex: /^\$bcrypt-sha256\$(2[axy]|2)\,[0-9]+\$[a-z0-9\/\.]{22}\$[a-z0-9\/\.]{31}$/i, hashes: ["BCrypt(SHA-256)"]},
            {regex: /^[a-f0-9]{32}:.{3}$/i, hashes: ["vBulletin < v3.8.5"]},
            {regex: /^[a-f0-9]{32}:.{30}$/i, hashes: ["vBulletin ≥ v3.8.5"]},
            {regex: /^[a-f0-9]{64}$/i, hashes: ["SHA-256","RIPEMD-256","Haval-256","Snefru-256","GOST R 34.11-94","SHA3-256","Skein-256","Skein-512(256)","Ventrilo"]},
            {regex: /^[a-f0-9]{32}:[a-z0-9]{32}$/i, hashes: ["Joomla < v2.5.18"]},
            {regex: /^[a-f-0-9]{32}:[a-f-0-9]{32}$/i, hashes: ["SAM(LM_Hash:NT_Hash)"]},
            {regex: /^[a-f0-9]{32}:[0-9]{32}:[0-9]{2}$/i, hashes: ["MD5(Chap)","iSCSI CHAP Authentication"]},
            {regex: /^\$episerver\$\*0\*[a-z0-9=\*\+]{52}$/i, hashes: ["EPiServer 6.x < v4"]},
            {regex: /^{ssha256}[a-z0-9\.\$]{63}$/i, hashes: ["AIX(ssha256)"]},
            {regex: /^[a-f0-9]{80}$/i, hashes: ["RIPEMD-320"]},
            {regex: /^\$episerver\$\*1\*[a-z0-9=\*\+]{68}$/i, hashes: ["EPiServer 6.x ≥ v4"]},
            {regex: /^0x0100[a-f0-9]{88}$/i, hashes: ["MSSQL(2000)"]},
            {regex: /^[a-f0-9]{96}$/i, hashes: ["SHA-384","SHA3-384","Skein-512(384)","Skein-1024(384)"]},
            {regex: /^{SSHA512}[a-z0-9\+\/]{96}={0,2}$/i, hashes: ["SSHA-512(Base64)","LDAP(SSHA-512)"]},
            {regex: /^{ssha512}[0-9]{2}\$[a-z0-9\.\/]{16,48}\$[a-z0-9\.\/]{86}$/i, hashes: ["AIX(ssha512)"]},
            {regex: /^[a-f0-9]{128}$/i, hashes: ["SHA-512","Whirlpool","Salsa10","Salsa20","SHA3-512","Skein-512","Skein-1024(512)"]},
            {regex: /^[a-f0-9]{136}$/i, hashes: ["OSX v10.7"]},
            {regex: /^0x0200[a-f0-9]{136}$/i, hashes: ["MSSQL(2012)"]},
            {regex: /^\$ml\$[0-9]+\$[a-f0-9]{64}\$[a-f0-9]{128}$/i, hashes: ["OSX v10.8","OSX v10.9"]},
            {regex: /^[a-f0-9]{256}$/i, hashes: ["Skein-1024"]},
            {regex: /^grub\.pbkdf2\.sha512\.[0-9]+\.[a-f0-9]+\.[a-f0-9]+$/i, hashes: ["GRUB 2"]},
            {regex: /^sha1\$[a-z0-9\/\.]{1,12}\$[a-f0-9]{40}$/i, hashes: ["Django(SHA-1)"]},
            {regex: /^[a-f0-9]{49}$/i, hashes: ["Citrix Netscaler"]},
            {regex: /^\$S\$[a-z0-9\/\.]{52}$/i, hashes: ["Drupal ≥ v7.x"]},
            {regex: /^\$5\$(rounds=[0-9]+\$)?[a-z0-9\/\.]{0,16}\$[a-z0-9\/\.]{43}$/i, hashes: ["SHA-256 Crypt"]},
            {regex: /^0x[a-f0-9]{4}[a-f0-9]{16}[a-f0-9]{64}$/i, hashes: ["Sybase ASE"]},
            {regex: /^\$6\$(rounds=[0-9]+\$)?[a-z0-9\/\.]{0,16}\$[a-z0-9\/\.]{86}$/i, hashes: ["SHA-512 Crypt"]},
            {regex: /^\$sha\$[a-z0-9]{1,16}\$([a-f0-9]{32}|[a-f0-9]{40}|[a-f0-9]{64}|[a-f0-9]{128}|[a-f0-9]{140})$/i, hashes: ["Minecraft(AuthMe Reloaded)"]},
            {regex: /^sha256\$[a-z0-9\/\.]{1,12}\$[a-f0-9]{64}$/i, hashes: ["Django(SHA-256)"]},
            {regex: /^sha384\$[a-z0-9\/\.]{1,12}\$[a-f0-9]{96}$/i, hashes: ["Django(SHA-384)"]},
            {regex: /^crypt1:[a-z0-9\+\=]{12}:[a-z0-9\+\=]{12}$/i, hashes: ["Clavister Secure Gateway"]},
            {regex: /^[a-f0-9]{112}$/i, hashes: ["Cisco VPN Client(PCF-File)"]},
            {regex: /^[a-f0-9]{1329}$/i, hashes: ["Microsoft MSTSC(RDP-File)"]},
            {regex: /^[^\\\/\:\*\?\"<\>\|]{1,20}::[^\\\/\:\*\?\"<\>\|]{1,20}:[a-f0-9]{48}:[a-f0-9]{48}:[a-f0-9]{16}$/i, hashes: ["NetNTLMv1-VANILLA / NetNTLMv1+ESS"]},
            {regex: /^[^\\\/\:\*\?\"<\>\|]{1,20}::[^\\\/\:\*\?\"<\>\|]{1,20}:[a-f0-9]{16}:[a-f0-9]{32}:[a-f0-9]+$/i, hashes: ["NetNTLMv2"]},
            {regex: /^\$krb5pa\$23\$user\$realm\$salt\$[a-f0-9]{104}$/i, hashes: ["Kerberos 5 AS-REQ Pre-Auth"]},
            {regex: /^\$scram\$[0-9]+\$[a-z0-9\/\.]{16}\$sha-1=[a-z0-9\/\.]{27},sha-256=[a-z0-9\/\.]{43},sha-512=[a-z0-9\/\.]{86}$/i, hashes: ["SCRAM Hash"]},
            {regex: /^[a-f0-9]{40}:[a-f0-9]{0,32}$/i, hashes: ["Redmine Project Management Web App"]},
            {regex: /^([0-9]{12})?\$[a-f0-9]{16}$/i, hashes: ["SAP CODVN B (BCODE)"]},
            {regex: /^([0-9]{12})?\$[a-f0-9]{40}$/i, hashes: ["SAP CODVN F/G (PASSCODE)"]},
            {regex: /^[a-z0-9\/\.]{30}(:.+)?$/i, hashes: ["Juniper Netscreen/SSG(ScreenOS)"]},
            {regex: /^0x[a-f0-9]{60}\s0x[a-f0-9]{40}$/i, hashes: ["EPi"]},
            {regex: /^[a-f0-9]{40}:[^*]{1,25}$/i, hashes: ["SMF ≥ v1.1"]},
            {regex: /^[a-f0-9]{40}:[a-f0-9]{40}$/i, hashes: ["Woltlab Burning Board 3.x"]},
            {regex: /^[a-f0-9]{130}(:[a-f0-9]{40})?$/i, hashes: ["IPMI2 RAKP HMAC-SHA1"]},
            {regex: /^[a-f0-9]{32}:[0-9]+:[a-z0-9_.+-]+@[a-z0-9-]+\.[a-z0-9-.]+$/i, hashes: ["Lastpass"]},
            {regex: /^[a-z0-9\/\.]{16}(:.{1,})?$/i, hashes: ["Cisco-ASA(MD5)"]},
            {regex: /^\$vnc\$\*[a-f0-9]{32}\*[a-f0-9]{32}$/i, hashes: ["VNC"]},
            {regex: /^[a-z0-9]{32}(:([a-z0-9-]+\.)?[a-z0-9-.]+\.[a-z]{2,7}:.+:[0-9]+)?$/i, hashes: ["DNSSEC(NSEC3)"]},
            {regex: /^(user-.+:)?\$racf\$\*.+\*[a-f0-9]{16}$/i, hashes: ["RACF"]},
            {regex: /^\$3\$\$[a-f0-9]{32}$/i, hashes: ["NTHash(FreeBSD Variant)"]},
            {regex: /^\$sha1\$[0-9]+\$[a-z0-9\/\.]{0,64}\$[a-z0-9\/\.]{28}$/i, hashes: ["SHA-1 Crypt"]},
            {regex: /^[a-f0-9]{70}$/i, hashes: ["hMailServer"]},
            {regex: /^[:\$][AB][:\$]([a-f0-9]{1,8}[:\$])?[a-f0-9]{32}$/i, hashes: ["MediaWiki"]},
            {regex: /^[a-f0-9]{140}$/i, hashes: ["Minecraft(xAuth)"]},
            {regex: /^\$pbkdf2-sha(1|256|512)\$[0-9]+\$[a-z0-9\/\.]{22}\$([a-z0-9\/\.]{27}|[a-z0-9\/\.]{43}|[a-z0-9\/\.]{86})$/i, hashes: ["PBKDF2(Generic)"]},
            {regex: /^\$p5k2\$[0-9]+\$[a-z0-9\/+=-]+\$[a-z0-9\/\+=-]{28}$/i, hashes: ["PBKDF2(Cryptacular)"]},
            {regex: /^\$p5k2\$[0-9]+\$[a-z0-9\/\.]+\$[a-z0-9\/\.]{32}$/i, hashes: ["PBKDF2(Dwayne Litzenberger)"]},
            {regex: /^{FSHP[0123]\|[0-9]+\|[0-9]+}[a-z0-9\/+=]+$/i, hashes: ["Fairly Secure Hashed Password"]},
            {regex: /^\$PHPS\$.+\$[a-f0-9]{32}$/i, hashes: ["PHPS"]},
            {regex: /^[0-9]{4}:[a-f0-9]{16}:[a-f0-9]{2080}$/i, hashes: ["1Password(Agile Keychain)"]},
            {regex: /^[a-f0-9]{64}:[a-f0-9]{32}:[0-9]{5}:[a-f0-9]{608}$/i, hashes: ["1Password(Cloud Keychain)"]},
            {regex: /^[a-f0-9]{256}:[a-f0-9]{256}:[a-f0-9]{16}:[a-f0-9]{16}:[a-f0-9]{320}:[a-f0-9]{16}:[a-f0-9]{40}:[a-f0-9]{40}:[a-f0-9]{32}$/i, hashes: ["IKE-PSK MD5"]},
            {regex: /^[a-f0-9]{256}:[a-f0-9]{256}:[a-f0-9]{16}:[a-f0-9]{16}:[a-f0-9]{320}:[a-f0-9]{16}:[a-f0-9]{40}:[a-f0-9]{40}:[a-f0-9]{40}$/i, hashes: ["IKE-PSK SHA1"]},
            {regex: /^[a-z0-9+\/]{27}=$/i, hashes: ["PeopleSoft"]},
            {regex: /^crypt\$[a-f0-9]{5}\$[a-z0-9.\/]{13}$/i, hashes: ["Django(DES Crypt Wrapper)"]},
            {regex: /^pbkdf2_sha256\$[0-9]+\$[a-z0-9]{1,}\$[a-z0-9+\/]{43}=$/i, hashes: ["Django(PBKDF2-HMAC-SHA256)"]},
            {regex: /^pbkdf2_sha1\$[0-9]+\$[a-z0-9]{1,}\$[a-z0-9+\/]{27}=$/i, hashes: ["Django(PBKDF2-HMAC-SHA1)"]},
            {regex: /^bcrypt(\$2[axy]|\$2)\$[0-9]{0,2}?\$[a-z0-9\/\.]{53}$/i, hashes: ["Django(BCrypt)"]},
            {regex: /^md5\$[a-f0-9]{1,}\$[a-f0-9]{32}$/i, hashes: ["Django(MD5)"]},
            {regex: /^\{PKCS5S2\}[a-z0-9+\/]{64}$/i, hashes: ["PBKDF2(Atlassian)"]},
            {regex: /^md5[a-f0-9]{32}$/i, hashes: ["PostgreSQL MD5"]},
            {regex: /^[a-f0-9]{32}(:.+)$/i, hashes: ["Skype"]},
            {regex: /^\([a-z0-9\/+]{49}\)$/i, hashes: ["Lotus Notes/Domino 8"]},
            {regex: /^SCRYPT:[0-9]{5}:[0-9]{1}:[0-9]{1}:[a-z0-9\/+]{14}==:[a-z0-9\/+]{43}=$/i, hashes: ["scrypt"]}
        ];
        var result = [];
        for(var hash_id=0; hash_id < hash_arr.length; hash_id++){
            var curr_hash = hash_arr[hash_id];
            
            //try to find matches
            if (phash.match(curr_hash.regex)){
                result = result.concat(curr_hash.hashes);
            }
        }
        if(result.length > 0){
            return result;
        }
        // else
        return new Error("Did not find any matches.");
    };
})(typeof exports === 'undefined'? this['hashID']={}: exports);
